bindings/matrix-sdk-ffi/src/timeline/mod.rs

```diff
@@ -132,24 +132,20 @@ impl Timeline {
 
 #[uniffi::export(async_runtime = "tokio")]
 impl Timeline {
-    pub async fn add_listener(
-        &self,
-        listener: Box<dyn TimelineListener>,
-    ) -> RoomTimelineListenerResult {
+    pub async fn add_listener(&self, listener: Box<dyn TimelineListener>) -> Arc<TaskHandle> {
         let (timeline_items, timeline_stream) = self.inner.subscribe_batched().await;
-        let timeline_stream = TaskHandle::new(RUNTIME.spawn(async move {
+
+        Arc::new(TaskHandle::new(RUNTIME.spawn(async move {
             pin_mut!(timeline_stream);
 
+            let initial_items = timeline_items.into_iter().map(TimelineItem::from_arc).collect();
+            listener.on_initial_items(initial_items);
```

Comment:
Simply call on_update here would be similar, and it removes the race condition on the consumer side. Otherwise it's still possible to get on_update actually called before on_initial_items on the host/consumer side depending on how busy the runtime is I guess. By only calling/using on_update, we remove a lot of complexity on the host side, and fix the race condition.

Reply:
How would that work? The two delegate methods have different signatures 🤔

Reply:
> it's still possible to get on_update actually called before on_initial_items on the host/consumer side depending on how busy the runtime is

Is this true? I'd have thought calling `on_initial_items` here would block until the consumer code had finished running?

Reply:
> How would that work? The two delegate methods have different signatures 🤔

You can wrap the initial values in a `TimelineDiff::Reset` or `TimelineDiff::Append`. Thoughts?

Reply:
> but on the consumer side

I don't buy this argument. This PR is fixing a real issue that's caused in the code we control. Fixing issues on the consumer side is outside our control, and while we can do our best to *limit* or *mitigate* damage by providing good APIs in the first place, this argument doesn't apply to the problem at hand here, IMO; if a consumer goes async-wild on their side, I'd rather consider it's on them and they might be shooting _themselves_ in the feet.

> You can wrap the initial values in a TimelineDiff::Reset or TimelineDiff::Append. Thoughts?

Sounds like a bit of API abuse to me, but I'll happily get rid of one new method to limit the API churn, so I'll consider this, thanks.

Reply:
It's actually great, thanks for the suggestion.